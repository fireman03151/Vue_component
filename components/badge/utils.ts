import { PresetColorTypes } from '../_util/colors';

export function isPresetColor(color?: string): boolean {
  return (PresetColorTypes as string[]).indexOf(color) !== -1;
}
