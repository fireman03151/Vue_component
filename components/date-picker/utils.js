export function formatDate(value, format) {
  const isFunction = function(obj) {
    return !!(obj && obj.constructor && obj.call && obj.apply);
  };
  
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }
  if (isFunction(format)) {
    const result = format(value);
    if (typeof result === 'string') {
      return result;
    } else {
      throw new Error('The function of format does not return a string');
    }
  }
  return value.format(format);
}
