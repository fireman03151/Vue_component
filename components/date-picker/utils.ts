import moment from 'moment';

type Value = moment.Moment | undefined | null;
type Format = string | string[] | undefined | ((val?: Value) => string | string[] | undefined);
export function formatDate(value: Value, format: Format) {
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }
  if (typeof format === 'function') {
    return format(value);
  }
  return value.format(format);
}
