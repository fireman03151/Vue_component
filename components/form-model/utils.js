export function getValueFromEvent(e) {
  // To support custom element
  if (!e || !e.target) {
    return e;
  }
  const { target } = e;
  return target.type === 'checkbox' ? target.checked : target.value;
}

export function getErrorStrs(errors) {
  if (errors) {
    return errors.map(e => {
      if (e && e.message) {
        return e.message;
      }
      return e;
    });
  }
  return errors;
}

export function getParams(ns, opt, cb) {
  let names = ns;
  let options = opt;
  let callback = cb;
  if (typeof names === 'string') {
    names = [names];
  }
  if (cb === undefined) {
    if (typeof names === 'function') {
      callback = names;
      options = {};
      names = undefined;
    } else if (Array.isArray(names)) {
      if (typeof options === 'function') {
        callback = options;
        options = {};
      } else {
        options = options || {};
      }
    } else {
      callback = options;
      options = names || {};
      names = undefined;
    }
  }
  return {
    names,
    options,
    callback,
  };
}

export function hasRules(validate) {
  if (validate) {
    return validate.some(item => {
      return item.rules && item.rules.length;
    });
  }
  return false;
}

export function computedStyle(el, prop) {
  const getComputedStyle = window.getComputedStyle;
  const style =
    // If we have getComputedStyle
    getComputedStyle
      ? // Query it
        // TODO: From CSS-Query notes, we might need (node, null) for FF
        getComputedStyle(el)
      : // Otherwise, we are in IE and use currentStyle
        el.currentStyle;
  if (style) {
    return style[
      // Switch to camelCase for CSSOM
      // DEV: Grabbed from jQuery
      // https://github.com/jquery/jquery/blob/1.9-stable/src/css.js#L191-L194
      // https://github.com/jquery/jquery/blob/1.9-stable/src/core.js#L593-L597
      prop.replace(/-(\w)/gi, (word, letter) => {
        return letter.toUpperCase();
      })
    ];
  }
  return undefined;
}

export function getScrollableContainer(n) {
  let node = n;
  let nodeName;
  /* eslint no-cond-assign:0 */
  while ((nodeName = node.nodeName.toLowerCase()) !== 'body') {
    const overflowY = computedStyle(node, 'overflowY');
    // https://stackoverflow.com/a/36900407/3040605
    if (
      node !== n &&
      (overflowY === 'auto' || overflowY === 'scroll') &&
      node.scrollHeight > node.clientHeight
    ) {
      return node;
    }
    node = node.parentNode;
  }
  return nodeName === 'body' ? node.ownerDocument : node;
}

// export async function finishOnAllFailed(rulePromises) {
//   return Promise.all(rulePromises).then(errorsList => {
//     const errors = [].concat(...errorsList);

//     return errors;
//   });
// }

// export async function finishOnFirstFailed(rulePromises) {
//   let count = 0;

//   return new Promise(resolve => {
//     rulePromises.forEach(promise => {
//       promise.then(errors => {
//         if (errors.length) {
//           resolve(errors);
//         }

//         count += 1;
//         if (count === rulePromises.length) {
//           resolve([]);
//         }
//       });
//     });
//   });
// }

// export function toArray(value) {
//   if (value === undefined || value === null) {
//     return [];
//   }

//   return Array.isArray(value) ? value : [value];
// }

// export function getNamePath(path) {
//   return toArray(path);
// }
