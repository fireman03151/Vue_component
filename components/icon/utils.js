import warning from '../_util/warning';

// These props make sure that the SVG behaviours like general text.
// Reference: https://blog.prototypr.io/align-svg-icons-to-text-and-say-goodbye-to-font-icons-d44b3d7b26b4
export const svgBaseProps = {
  width: '1em',
  height: '1em',
  fill: 'currentColor',
  'aria-hidden': 'true',
  focusable: 'false',
};

const fillTester = /-fill$/;
const outlineTester = /-o$/;
const twoToneTester = /-twotone$/;

export function getThemeFromTypeName(type) {
  let result = null;
  if (fillTester.test(type)) {
    result = 'filled';
  } else if (outlineTester.test(type)) {
    result = 'outlined';
  } else if (twoToneTester.test(type)) {
    result = 'twoTone';
  }
  return result;
}

export function removeTypeTheme(type) {
  return type
    .replace(fillTester, '')
    .replace(outlineTester, '')
    .replace(twoToneTester, '');
}

export function withThemeSuffix(type, theme) {
  let result = type;
  if (theme === 'filled') {
    result += '-fill';
  } else if (theme === 'outlined') {
    result += '-o';
  } else if (theme === 'twoTone') {
    result += '-twotone';
  } else {
    warning(false, `This icon '${type}' has unknown theme '${theme}'`);
  }
  return result;
}

// For alias or compatibility
export function alias(type) {
  switch (type) {
    case 'cross':
      return 'close';
    default:
  }
  return type;
}
