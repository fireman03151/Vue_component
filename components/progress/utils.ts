export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}
