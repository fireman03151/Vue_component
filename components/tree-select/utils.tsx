import { flattenChildren, isValidElement } from '../_util/props-util';

export function convertChildrenToData(nodes: any[]): any[] {
  return flattenChildren(nodes)
    .map(node => {
      if (!isValidElement(node) || !node.type) {
        return null;
      }
      const { default: d, ...restSlot } = node.children || {};
      const children = d ? d() : [];
      const {
        key,
        props: { value, ...restProps },
      } = node;

      const data = {
        key,
        value,
        ...restProps,
      };
      Object.keys(restSlot).forEach(p => {
        if (typeof restSlot[p] === 'function') {
          data[p] = <>{restSlot[p]()}</>;
        }
      });
      const childData = convertChildrenToData(children);
      if (childData.length) {
        data.children = childData;
      }

      return data;
    })
    .filter(data => data);
}
