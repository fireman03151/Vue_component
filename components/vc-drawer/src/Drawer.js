import classnames from 'classnames'
import Vue from 'vue'
import BaseMixin from '../../_util/BaseMixin'
import antRefDirective from '../../_util/antRefDirective'
import { initDefaultProps, getEvents } from '../../_util/props-util'
import { cloneElement } from '../../_util/vnode'
import ContainerRender from '../../_util/ContainerRender'
import getScrollBarSize from '../../_util/getScrollBarSize'
import drawerProps from './drawerProps'
import {
  dataToArray,
  transitionEnd,
  trnasitionStr,
  addEventListener,
  removeEventListener,
  transformArguments,
  isNumeric,
} from './utils'

function noop () {}

const currentDrawer = {}
const windowIsUndefined = typeof window === 'undefined'
Vue.use(antRefDirective)
const Drawer = {
  mixins: [BaseMixin],
  props: initDefaultProps(drawerProps, {
    prefixCls: 'drawer',
    placement: 'left',
    getContainer: 'body',
    level: 'all',
    duration: '.3s',
    ease: 'cubic-bezier(0.78, 0.14, 0.15, 0.86)',
    firstEnter: false, // 记录首次进入.
    showMask: true,
    handler: true,
    maskStyle: {},
    wrapClassName: '',
  }),
  data () {
    this.levelDom = []
    this.contentDom = null
    this.maskDom = null
    this.handlerdom = null
    this.mousePos = null
    this.sFirstEnter = this.firstEnter
    this.timeout = null
    this.children = null
    this.drawerId = Number((Date.now() + Math.random()).toString()
      .replace('.', Math.round(Math.random() * 9))).toString(16)
    const open = this.open !== undefined ? this.open : !!this.defaultOpen
    currentDrawer[this.drawerId] = open
    this.orignalOpen = this.open
    this.preProps = { ...this.$props }
    return {
      sOpen: open,
    }
  },
  mounted () {
    this.$nextTick(() => {
      if (!windowIsUndefined) {
        let passiveSupported = false
        window.addEventListener(
          'test',
          null,
          Object.defineProperty({}, 'passive', {
            get: () => {
              passiveSupported = true
              return null
            },
          })
        )
        this.passive = passiveSupported ? { passive: false } : false
      }
      const open = this.getOpen()
      if (this.handler || open || this.sFirstEnter) {
        this.getDefault(this.$props)
        if (open) {
          this.isOpenChange = true
        }
        this.$forceUpdate()
      }
    })
  },
  watch: {
    open (val) {
      if (val !== undefined && val !== this.preProps.open) {
        this.isOpenChange = true
        // 没渲染 dom 时，获取默认数据;
        if (!this.container) {
          this.getDefault(this.$props)
        }
        this.setState({
          sOpen: open,
        })
      }
      this.preProps.open = val
    },
    placement (val) {
      if (val !== this.preProps.placement) {
        // test 的 bug, 有动画过场，删除 dom
        this.contentDom = null
      }
      this.preProps.placement = val
    },
    level (val) {
      if (this.preProps.level !== val) {
        this.getParentAndLevelDom(this.$props)
      }
      this.preProps.level = val
    },
  },
  updated () {
    this.$nextTick(() => {
      // dom 没渲染时，重走一遍。
      if (!this.sFirstEnter && this.container) {
        this.$forceUpdate()
        this.sFirstEnter = true
      }
    })
  },
  beforeDestroy () {
    delete currentDrawer[this.drawerId]
    delete this.isOpenChange
    if (this.container) {
      if (this.sOpen) {
        this.setLevelDomTransform(false, true)
      }
      document.body.style.overflow = ''
      // 拦不住。。直接删除；
      if (this.getSelfContainer) {
        this.container.parentNode.removeChild(this.container)
      }
    }
    this.sFirstEnter = false
    clearTimeout(this.timeout)
    // 需要 didmount 后也会渲染，直接 unmount 将不会渲染，加上判断.
    if (this.renderComponent) {
      this.renderComponent({
        afterClose: this.removeContainer,
        onClose () { },
        visible: false,
      })
    }
  },
  methods: {
    onMaskTouchEnd (e) {
      this.$emit('maskClick', e)
      this.onTouchEnd(e, true)
    },
    onIconTouchEnd (e) {
      this.$emit('handleClick', e)
      this.onTouchEnd(e)
    },
    onTouchEnd (e, close) {
      if (this.open !== undefined) {
        return
      }
      const open = close || this.sOpen
      this.isOpenChange = true
      this.setState({
        sOpen: !open,
      })
    },
    onWrapperTransitionEnd (e) {
      if (e.target === this.contentWrapper) {
        this.dom.style.transition = ''
        if (!this.sOpen && this.getCrrentDrawerSome()) {
          document.body.style.overflowX = ''
          if (this.maskDom) {
            this.maskDom.style.left = ''
            this.maskDom.style.width = ''
          }
        }
      }
    },
    getDefault (props) {
      this.getParentAndLevelDom(props)
      if (props.getContainer || props.parent) {
        this.container = this.defaultGetContainer()
      }
    },
    getCrrentDrawerSome () {
      return !Object.keys(currentDrawer).some(key => currentDrawer[key])
    },
    getSelfContainer () {
      return this.container
    },
    getParentAndLevelDom (props) {
      if (windowIsUndefined) {
        return
      }
      const { level, getContainer } = props
      this.levelDom = []
      if (getContainer) {
        if (typeof getContainer === 'string') {
          const dom = document.querySelectorAll(getContainer)[0]
          this.parent = dom
        }
        if (typeof getContainer === 'function') {
          this.parent = getContainer()
        }
        if (typeof getContainer === 'object' && getContainer instanceof window.HTMLElement) {
          this.parent = getContainer
        }
      }
      if (!getContainer && this.container) {
        this.parent = this.container.parentNode
      }
      if (level === 'all') {
        const children = Array.prototype.slice.call(this.parent.children)
        children.forEach(child => {
          if (child.nodeName !== 'SCRIPT' && child.nodeName !== 'STYLE' &&
            child.nodeName !== 'LINK' && child !== this.container) {
            this.levelDom.push(child)
          }
        })
      } else if (level) {
        dataToArray(level).forEach(key => {
          document.querySelectorAll(key).forEach(item => {
            this.levelDom.push(item)
          })
        })
      }
    },
    setLevelDomTransform (open, openTransition, placementName, value) {
      const { placement, levelMove, duration, ease } = this.$props
      if (!windowIsUndefined) {
        this.levelDom.forEach(dom => {
          if (this.isOpenChange || openTransition) {
            /* eslint no-param-reassign: "error" */
            dom.style.transition = `transform ${duration} ${ease}`
            addEventListener(dom, transitionEnd, this.trnasitionEnd)
            let levelValue = open ? value : 0
            if (levelMove) {
              const $levelMove = transformArguments(levelMove, { target: dom, open })
              levelValue = open ? $levelMove[0] : $levelMove[1] || 0
            }
            const $value = typeof levelValue === 'number' ? `${levelValue}px` : levelValue
            const placementPos = placement === 'left' || placement === 'top' ? $value : `-${$value}`
            dom.style.transform = levelValue ? `${placementName}(${placementPos})` : ''
            dom.style.msTransform = levelValue ? `${placementName}(${placementPos})` : ''
          }
        })
        // 处理 body 滚动
        const eventArray = ['touchstart']
        const domArray = [document.body, this.maskDom, this.handlerdom, this.contentDom]
        const right = getScrollBarSize(1)
        let widthTransition = `width ${duration} ${ease}`
        const trannsformTransition = `transform ${duration} ${ease}`
        if (open && document.body.style.overflow !== 'hidden') {
          document.body.style.overflow = 'hidden'
          if (right) {
            document.body.style.position = 'relative'
            document.body.style.width = `calc(100% - ${right}px)`
            this.dom.style.transition = 'none'
            switch (placement) {
              case 'right':
                this.dom.style.transform = `translateX(-${right}px)`
                this.dom.style.msTransform = `translateX(-${right}px)`
                break
              case 'top':
              case 'bottom':
                this.dom.style.width = `calc(100% - ${right}px)`
                this.dom.style.transform = 'translateZ(0)'
                break
              default:
                break
            }
            clearTimeout(this.timeout)
            this.timeout = setTimeout(() => {
              this.dom.style.transition = `${trannsformTransition},${widthTransition}`
              this.dom.style.width = ''
              this.dom.style.transform = ''
              this.dom.style.msTransform = ''
            })
          }
          // 手机禁滚
          domArray.forEach((item, i) => {
            if (!item) {
              return
            }
            addEventListener(
              item,
              eventArray[i] || 'touchmove',
              i ? this.removeMoveHandler : this.removeStartHandler,
              this.passive
            )
          })
        } else if (this.getCrrentDrawerSome()) {
          document.body.style.overflow = ''
          if ((this.isOpenChange || openTransition) && right) {
            document.body.style.position = ''
            document.body.style.width = ''
            if (trnasitionStr) {
              document.body.style.overflowX = 'hidden'
            }
            this.dom.style.transition = 'none'
            let heightTransition
            switch (placement) {
              case 'right': {
                this.dom.style.transform = `translateX(${right}px)`
                this.dom.style.msTransform = `translateX(${right}px)`
                this.dom.style.width = '100%'
                widthTransition = `width 0s ${ease} ${duration}`
                if (this.maskDom) {
                  this.maskDom.style.left = `-${right}px`
                  this.maskDom.style.width = `calc(100% + ${right}px)`
                }
                break
              }
              case 'top':
              case 'bottom': {
                this.dom.style.width = `calc(100% + ${right}px)`
                this.dom.style.height = '100%'
                this.dom.style.transform = 'translateZ(0)'
                heightTransition = `height 0s ${ease} ${duration}`
                break
              }
              default:
                break
            }
            clearTimeout(this.timeout)
            this.timeout = setTimeout(() => {
              this.dom.style.transition = `${trannsformTransition},${
                heightTransition ? `${heightTransition},` : ''}${widthTransition}`
              this.dom.style.transform = ''
              this.dom.style.msTransform = ''
              this.dom.style.width = ''
              this.dom.style.height = ''
            })
          }
          domArray.forEach((item, i) => {
            if (!item) {
              return
            }
            removeEventListener(
              item,
              eventArray[i] || 'touchmove',
              i ? this.removeMoveHandler : this.removeStartHandler,
              this.passive
            )
          })
        }
      }
      const { change } = this.$listeners
      if (change && this.isOpenChange && this.sFirstEnter) {
        change(open)
        this.isOpenChange = false
      }
    },
    getChildToRender (open) {
      const {
        prefixCls,
        placement,
        handler,
        showMask,
        maskStyle,
        width,
        height,
        wrapClassName,
      } = this.$props
      const children = this.$slots.default
      const wrapperClassname = classnames(prefixCls, {
        [`${prefixCls}-${placement}`]: true,
        [`${prefixCls}-open`]: open,
        [wrapClassName]: !!wrapClassName,
      })
      const isOpenChange = this.isOpenChange
      const isHorizontal = placement === 'left' || placement === 'right'
      const placementName = `translate${isHorizontal ? 'X' : 'Y'}`
      // 百分比与像素动画不同步，第一次打用后全用像素动画。
      // const defaultValue = !this.contentDom || !level ? '100%' : `${value}px`;
      const placementPos =
        placement === 'left' || placement === 'top' ? '-100%' : '100%'
      const transform = open ? '' : `${placementName}(${placementPos})`
      if (isOpenChange === undefined || isOpenChange) {
        const contentValue = this.contentDom ? this.contentDom.getBoundingClientRect()[
          isHorizontal ? 'width' : 'height'
        ] : 0
        const value = (isHorizontal ? width : height) || contentValue
        this.setLevelDomTransform(open, false, placementName, value)
      }
      let handlerCildren
      if (handler !== false) {
        const handlerDefalut = (
          <div class='drawer-handle'>
            <i class='drawer-handle-icon' />
          </div>
        )
        const { handler: handlerSlot } = this.$slots
        const handlerSlotVnode = handlerSlot || handlerDefalut
        const { click: handleIconClick } = getEvents(handlerSlotVnode)
        handlerCildren = cloneElement(handlerSlotVnode, {
          on: {
            click: (e) => {
              handleIconClick && handleIconClick()
              this.onIconTouchEnd(e)
            },
          },
          directives: [{
            name: 'ant-ref',
            value: (c) => {
              this.handlerdom = c
            },
          }],
        })
      }
      const domContProps = {
        class: wrapperClassname,
        directives: [{
          name: 'ant-ref',
          value: (c) => {
            this.dom = c
          },
        }],
        on: {
          transitionend: this.onWrapperTransitionEnd,
        },
      }
      const directivesMaskDom = [{
        name: 'ant-ref',
        value: (c) => {
          this.maskDom = c
        },
      }]
      const directivesContentWrapper = [{
        name: 'ant-ref',
        value: (c) => {
          this.contentWrapper = c
        },
      }]
      const directivesContentDom = [{
        name: 'ant-ref',
        value: (c) => {
          this.contentDom = c
        },
      }]
      return (
        <div
          {...domContProps}
        >
          {showMask && (
            <div
              class={`${prefixCls}-mask`}
              onClick={this.onMaskTouchEnd}
              style={maskStyle}
              {...{ directives: directivesMaskDom }}
            />
          )}
          <div
            class={`${prefixCls}-content-wrapper`}
            style={{
              transform,
              msTransform: transform,
              width: isNumeric(width) ? `${width}px` : width,
              height: isNumeric(height) ? `${height}px` : height,
            }}
            {...{ directives: directivesContentWrapper }}
          >
            <div
              class={`${prefixCls}-content`}
              {...{ directives: directivesContentDom }}
              onTouchstart={open ? this.removeStartHandler : noop} // 跑用例用
              onTouchmove={open ? this.removeMoveHandler : noop} // 跑用例用
            >
              {children}
            </div>
            {handlerCildren}
          </div>
        </div>
      )
    },
    getOpen () {
      return this.open !== undefined ? this.open : this.sOpen
    },
    getTouchParentScroll (root, currentTarget, differX, differY) {
      /**
       * 增加 rect。
       * 当父级 dom 的 overflow 未开启滚动时，scrollLeft 或 scrollTop 为 0, 而 scrollWidth 增加了，
       * 父级是跟随子级的 rect, 直到父级设定了滚动.
       */
      const rect = currentTarget.getBoundingClientRect()
      if (!currentTarget) {
        return false
      } else if (
        (((currentTarget.scrollTop + currentTarget.offsetHeight + currentTarget.offsetTop >=
          currentTarget.scrollHeight + rect.top &&
          differY < 0) ||
          (currentTarget.scrollTop <= 0 && differY > 0)) &&
          Math.max(Math.abs(differX), Math.abs(differY)) === Math.abs(differY)) ||
        (((currentTarget.scrollLeft + currentTarget.offsetWidth + currentTarget.offsetLeft >=
          currentTarget.scrollWidth + rect.left &&
          differX < 0) ||
          (currentTarget.scrollLeft <= 0 && differX > 0)) &&
          Math.max(Math.abs(differX), Math.abs(differY)) === Math.abs(differX))
      ) {
        return root === currentTarget ||
          this.getTouchParentScroll(root, currentTarget.parentNode, differX, differY)
      }
      return false
    },
    removeStartHandler (e) {
      if (e.touches.length > 1) {
        return
      }
      this.startPos = {
        x: e.touches[0].clientX,
        y: e.touches[0].clientY,
      }
    },
    removeMoveHandler (e) {
      if (e.changedTouches.length > 1) {
        return
      }
      const currentTarget = e.currentTarget
      const differX = e.changedTouches[0].clientX - this.startPos.x
      const differY = e.changedTouches[0].clientY - this.startPos.y
      if (
        currentTarget === this.maskDom ||
        currentTarget === this.handlerdom ||
        currentTarget === this.contentDom &&
        this.getTouchParentScroll(currentTarget, e.target, differX, differY)
      ) {
        e.preventDefault()
      }
    },
    trnasitionEnd (e) {
      removeEventListener(e.target, transitionEnd, this.trnasitionEnd)
      e.target.style.transition = ''
    },
    defaultGetContainer () {
      if (windowIsUndefined) {
        return null
      }
      const container = document.createElement('div')
      this.parent.appendChild(container)
      if (this.wrapperClassName) {
        container.className = this.wrapperClassName
      }
      return container
    },
  },

  render () {
    const { getContainer, wrapperClassName } = this.$props
    const open = this.getOpen()
    currentDrawer[this.drawerId] = open ? this.container : open
    this.children = this.getChildToRender(this.sFirstEnter ? open : false)
    if (!getContainer) {
      const directives = [{
        name: 'ant-ref',
        value: (c) => {
          this.container = c
        },
      }]
      return (
        <div
          class={wrapperClassName}
          {...{ directives }}
        >
          {this.children}
        </div>
      )
    }
    if (!this.container || !open && !this.sFirstEnter) {
      return null
    }

    return (
      <ContainerRender
        parent={this}
        visible
        autoMount
        autoDestroy={false}
        getComponent={() => this.children}
        getContainer={this.getSelfContainer}
        children={({ renderComponent, removeContainer }) => {
          this.renderComponent = renderComponent
          this.removeContainer = removeContainer
          return null
        }}
      >
      </ContainerRender>
    )
  },
}

export default Drawer
