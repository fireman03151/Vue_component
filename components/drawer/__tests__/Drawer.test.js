import { mount } from '@vue/test-utils';
import { asyncExpect } from '@/tests/utils';
import Drawer from '..';

const DrawerCom = {
  props: {
    visible: {
      type: Boolean,
      default: false,
    },
    destroyOnClose: {
      type: Boolean,
      default: false,
    },
    getContainer: {
      type: Boolean,
      default: false,
    },
    wrapClassName: {
      type: String,
      default: '',
    },
  },
  render() {
    const drawerProps = {
      props: {
        destroyOnClose: true,
        getContainer: false,
        visible: false,
        wrapClassName: this.wrapClassName,
      },
    };
    return <Drawer {...drawerProps}>Here is content of Drawer</Drawer>;
  },
};

describe('Drawer', () => {
  it('render correctly', async () => {
    const props = {
      propsData: {
        visible: true,
        width: 400,
        getContainer: false,
      },
      slots: {
        default: ['Here is content of Drawer'],
      },
      sync: false,
    };
    const wrapper = mount(Drawer, props);
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('render top drawer', async () => {
    const props = {
      propsData: {
        visible: true,
        height: 400,
        placement: 'top',
        getContainer: false,
      },
      slots: {
        default: 'Here is content of Drawer',
      },
      sync: false,
    };
    const wrapper = mount(Drawer, props);
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('have a title', async () => {
    const props = {
      propsData: {
        visible: true,
        title: 'Test Title',
        getContainer: false,
      },
      slots: {
        default: 'Here is content of Drawer',
      },
      sync: false,
    };
    const wrapper = mount(Drawer, props);
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('closable is false', async () => {
    const props = {
      propsData: {
        visible: true,
        closable: false,
        getContainer: false,
      },
      slots: {
        default: 'Here is content of Drawer',
      },
      sync: false,
    };
    const wrapper = mount(Drawer, props);
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('destroyOnClose is true', async () => {
    const props = {
      propsData: {
        destroyOnClose: true,
        closable: false,
        getContainer: false,
      },
      slots: {
        default: 'Here is content of Drawer',
      },
      sync: false,
    };
    const wrapper = mount(Drawer, props);
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('class is test_drawer', async () => {
    const props = {
      propsData: {
        wrapClassName: 'test_drawer',
      },
      sync: false,
    };
    const wrapper = mount(DrawerCom, props);
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
