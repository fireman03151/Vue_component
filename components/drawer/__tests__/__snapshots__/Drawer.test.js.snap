// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer class is test_drawer 1`] = `
<div>
  <div class="ant-drawer ant-drawer-right test_drawer">
    <div class="ant-drawer-mask"></div>
    <div class="ant-drawer-content-wrapper" style="transform: translateX(100%); width: 256px;">
      <div class="ant-drawer-content">
        <div class="ant-drawer-wrapper-body" style="overflow: auto; height: 100%; opacity: 0;">
          <button aria-label="Close" class="ant-drawer-close"><span class="ant-drawer-close-x"></span></button>
          <div class="ant-drawer-body">Here is content of Drawer</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Drawer closable is false 1`] = `
<div>
  <div class="ant-drawer ant-drawer-right">
    <div class="ant-drawer-mask"></div>
    <div class="ant-drawer-content-wrapper" style="transform: translateX(100%); width: 256px;">
      <div class="ant-drawer-content">
        <div class="ant-drawer-wrapper-body" style="overflow: auto; height: 100%;">
          <div class="ant-drawer-body">Here is content of Drawer</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Drawer destroyOnClose is true 1`] = `
<div>
  <div class="ant-drawer ant-drawer-right">
    <div class="ant-drawer-mask"></div>
    <div class="ant-drawer-content-wrapper" style="transform: translateX(100%); width: 256px;">
      <div class="ant-drawer-content">
        <div class="ant-drawer-wrapper-body" style="overflow: auto; height: 100%; opacity: 0;">
          <div class="ant-drawer-body">Here is content of Drawer</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Drawer have a title 1`] = `
<div>
  <div class="ant-drawer ant-drawer-right">
    <div class="ant-drawer-mask"></div>
    <div class="ant-drawer-content-wrapper" style="transform: translateX(100%); width: 256px;">
      <div class="ant-drawer-content">
        <div class="ant-drawer-wrapper-body" style="overflow: auto; height: 100%;">
          <div class="ant-drawer-header">
            <div class="ant-drawer-title">Test Title</div>
          </div>
          <button aria-label="Close" class="ant-drawer-close"><span class="ant-drawer-close-x"></span></button>
          <div class="ant-drawer-body">Here is content of Drawer</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Drawer render correctly 1`] = `
<div>
  <div class="ant-drawer ant-drawer-right">
    <div class="ant-drawer-mask"></div>
    <div class="ant-drawer-content-wrapper" style="transform: translateX(100%); width: 400px;">
      <div class="ant-drawer-content">
        <div class="ant-drawer-wrapper-body" style="overflow: auto; height: 100%;">
          <button aria-label="Close" class="ant-drawer-close"><span class="ant-drawer-close-x"></span></button>
          <div class="ant-drawer-body">Here is content of Drawer</div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
